pilot/pkg/serviceregistry/kube/controller/discoverynamespacecontroller.go

```diff
@@ -0,0 +1,117 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"github.com/hashicorp/go-multierror"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube/controller/filter"
+	kubelib "istio.io/istio/pkg/kube"
+	"istio.io/pkg/log"
+)
+
+// On namespace membership change, request a full xDS update since services need to be recomputed
+func (c *Controller) initDiscoveryNamespaceHandlers(kubeClient kubelib.Client, endpointMode EndpointMode) {
+	otype := "Namespaces"
+	c.nsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			incrementEvent(otype, "add")
+			ns := obj.(*v1.Namespace)
+			if ns.Labels[filter.PilotDiscoveryLabelName] == filter.PilotDiscoveryLabelValue {
+				c.queue.Push(func() error {
+					// labeled namespace was newly created, issue creates for all services, pods, and endpoints in namespace
+					if err := c.SyncAll(); err != nil {
```

Comment:
syncAll may be very expensive. I don't think any of these resources have cross-namespace references, it may be possible to only sync the ones in that new namespace?